### PR TITLE
Fix compilation without libz

### PR DESCRIPTION
**Pull request by danielgustafsson**
Configuring the tree with --without-zlib generated compiler errors as
we didn't properly guard usage of zlib with the autoconf featuremacro
HAVE_LIBZ. This fixes the build by wrapping all callsites and provide
a dummy implementation to error out in as a fallback (as we do for the
other algorithms currently). While at it, compresstype_is_valid gets a
little bit of polish and C99 loop initialization.

While it will be rare that anyone wants to build explicitly without
libz support (the default is to include it), it should still work. It
is such a cornercase though that I don't see much point in adding a
test for it

### CODE REVIEW

File: src/backend/utils/misc/fstream/gfile.c [hunk 1]
```diff
@@ -32,6 +32,7 @@
 #include <fcntl.h>
 #include <sys/stat.h>
 #include <sys/file.h>   /* for flock */
+#include <unistd.h>
```
Comment (hlinnaka) [inline src/backend/utils/misc/fstream/gfile.c]: unrelated change?

Comment (danielgustafsson) [reply src/backend/utils/misc/fstream/gfile.c]: Actually not, `gfile.c` use `read()/write()` from `unistd.h` which get's sort of accidentally included via `zconf.h` in ZLib, so when building without ZLib the file doesn't compile.

**Comment by danielgustafsson** [comment]
Pushed to master, thanks for review!
